CrayPE fix for binutils

lukebroskop opened the pull request:
Allow for multiple definitions at link time (for CrayPE)

lukebroskop committed:
CrayPE fix for binutils

Allow for multiple definitions at link time (for CrayPE)

tldahlgren commented on var/spack/repos/builtin/packages/binutils/package.py:
```diff
@@ -71,6 +71,15 @@ class Binutils(AutotoolsPackage, GNUMirrorPackage):
     # --disable-ld flag
     conflicts('~ld', '+gold')
 
+    def setup_environment(self, spack_env, run_env):
```
Do you mean to use `setup_build_environment(self, env)` here?  The `setup_environment` method is a legacy function.

tldahlgren replied:
Or `setup_run_environment(self, env)`?

lukebroskop replied:
Thanks @tldahlgren , I would be adding these flags during the build step. However I did not see a reference to setup_build_environment (or setup_run_environment) in [https://spack-tutorial.readthedocs.io/en/latest/search.html?q=setup_build_environment&check_keywords=yes&area=default#](https://spack-tutorial.readthedocs.io/en/latest/search.html?q=setup_build_environment&check_keywords=yes&area=default#)

The current spack documentation indicates that setup_environment should be used: [https://spack-tutorial.readthedocs.io/en/latest/search.html?q=setup_build_environment&check_keywords=yes&area=default#](https://spack-tutorial.readthedocs.io/en/latest/search.html?q=setup_build_environment&check_keywords=yes&area=default#).

Perhaps the updated features are not documented yet or I'm looking at the wrong documents page?

tldahlgren replied:
> Thanks @tldahlgren , I would be adding these flags during the build step. However I did not see a reference to setup_build_environment (or setup_run_environment) in https://spack-tutorial.readthedocs.io/en/latest/search.html?q=setup_build_environment&check_keywords=yes&area=default#
> 
> The current spack documentation indicates that setup_environment should be used: https://spack-tutorial.readthedocs.io/en/latest/search.html?q=setup_build_environment&check_keywords=yes&area=default#.
> 
> Perhaps the updated features are not documented yet or I'm looking at the wrong documents page?

That's good to know, thanks!  There are *no* buillt-in packages using `setup_environment` anymore.  The tutorial is out-of-date and needs to be updated.

lukebroskop replied:
@tldahlgren thanks for the review, I've made your suggested change.

tldahlgren reviewed (changes requested):
Refer to previous questions/comments.

tldahlgren commented:
FWIW. I created an issue in the Spack tutorial (https://github.com/spack/spack-tutorial/issues/125) to address this problem.  Thanks for letting us know.

lukebroskop committed:
Update package.py

change legacy setup_environment function call for supported setup_build_environment function call

lukebroskop committed:
Update package.py

need to fix the setup_build_environment call

lukebroskop committed:
Update package.py

fix

tldahlgren commented on var/spack/repos/builtin/packages/binutils/package.py:
```diff
@@ -71,6 +71,14 @@ class Binutils(AutotoolsPackage, GNUMirrorPackage):
+    def setup_build_environment(self, env):
+
+        if self.spec.satisfies('%cce'):
+            setup_env.append_flags('LDFLAGS', '-Wl,-z,muldefs')
```
```suggestion
            env.append_flags('LDFLAGS', '-Wl,-z,muldefs')
```

tldahlgren commented on var/spack/repos/builtin/packages/binutils/package.py:
```diff
@@ -71,6 +71,14 @@ class Binutils(AutotoolsPackage, GNUMirrorPackage):
+            setup_env.append_flags('LDFLAGS', '-Wl,-z,muldefs')
+
+        if '+nls' in spec:
+            setup_env.append_flags('LDFLAGS', '-lintl')
```
```suggestion
            env.append_flags('LDFLAGS', '-lintl')
```

tldahlgren reviewed (changes requested):
Need to use the proper argument

lukebroskop committed:
Update var/spack/repos/builtin/packages/binutils/package.py

Co-authored-by: Tamara Dahlgren <35777542+tldahlgren@users.noreply.github.com>

lukebroskop committed:
Update var/spack/repos/builtin/packages/binutils/package.py

Co-authored-by: Tamara Dahlgren <35777542+tldahlgren@users.noreply.github.com>

haampie commented on var/spack/repos/builtin/packages/binutils/package.py:
```diff
@@ -99,7 +107,7 @@ def configure_args(self):
         if '+nls' in spec:
             args.append('--enable-nls')
-            args.append('LDFLAGS=-lintl')
+
```
```suggestion
```

lukebroskop committed:
Update var/spack/repos/builtin/packages/binutils/package.py

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>
